### DISCUSSION TITLE
TP6 - Por qué si `main` retorna un numero negativo, veo algo positivo (127)

### DISCUSSION THREAD

**Discussion by alengeor** [INGCOM-UNRN-P1/cursada-2023#173] [Preguntas y respuestas] [answered]
Buenas tardes!
Estoy comenzando con el nuevo TP, y me pasa algo rarisimo...
Si en mi codigo pretendo retornar -1, el echo me tira 127. Pero si en la misma funcion, bajo las mismas condiciones, quiero retornar por ejemplo 7, echo si me tira bien el numero. Van capturas

![image](https://github.com/INGCOM-UNRN-P1/cursada-2023/assets/110181852/b640d7ce-8012-40b2-b9f1-f2ff20615074)

No se si este es el lugar para copiar y pegar el codigo. Martín, si en la documentacion aclaro todos los posibles retornos, ¿puedo retornar con 7 en caso de que el orden sea descendente? Aparte de esto, todos los codigos de error funcionan perfecto.

Gracias!

**Comment by alengeor** [comment]
Aclaración: esto pasa con todos los retornos enteros negativos del programa. Misma situación, si cambio el numero de, por ejemplo, return -2 a return 2, funciona perfectamente.

Si no, el programa funciona pero echo tira 127.

**Reply by martinvilu** [reply, marked as answer]
La situación se da porque el retorno de la función, no va directamente al sistema operativo, sinó que pasa por la función `exit()`, https://en.cppreference.com/w/c/program/exit

La cual está pensada para recibir solo `EXIT_SUCCESS` y `EXIT_FAILURE`; que son positivos.

Algunas implementaciones de C, admiten que el retorno sea cualquier `int`, pero al `gcc` no le gusta mucho y lo que ves como `127`, es el `-1` que pasas, pero interpretado como un número sin signo de 8-bits (las cosas raras de C, porque un `int` es de cuatro veces más grande, con 32-bits)
```
binario      c/signo   s/signo
0000 0000 ->    0
0000 0001 ->    1
1000 0000 ->   -1   ->   127
```
El bit más a la izquierda, es usado para el signo, pero no deja de ser un valor binario más.

* solo el negativo tiene una interpretación diferente, porque los valores entre las posiciones 0 a 7 son 0.


Eeeeen resumen; retorná valores positivos.

**Reply by IoachimusRoderici** [reply]
> Eeeeen resumen; retorná valores positivos.

¿Y qué hacemos con `esta_ordenado`, donde la consigna pide retornar -1? ¿elegimos un código nuevo?¿dejamos que sea 127?

@martinvilu

**Comment by alengeor** [comment]
![thumbs-thumbs-up-kid](https://github.com/INGCOM-UNRN-P1/cursada-2023/assets/110181852/ecda24ad-a1ea-4991-a923-6a74111ee69b)
